src/utils.py

```diff
@@ -386,6 +386,107 @@ def powerset(seq: Sequence, exclude_empty: bool) -> Iterator[Sequence]:
                                          for r in range(start, len(seq)+1))
 
 
+_S = TypeVar('_S', bound=Hashable)  # state in heuristic search
+_A = TypeVar('_A')  # action in heuristic search
+
+
+@dataclass(frozen=True)
+class _HeuristicSearchNode(Generic[_S, _A]):
+    state: _S
+    edge_cost: float
+    cumulative_cost: float
+    parent: Optional[_HeuristicSearchNode[_S, _A]] = None
+    action: Optional[_A] = None
+
+
+def _run_heuristic_search(
+    initial_state: _S,
+    check_goal: Callable[[_S], bool],
+    get_successors: Callable[[_S], Iterator[Tuple[_A, _S, float]]],
+    get_priority: Callable[[_HeuristicSearchNode[_S, _A]], Any],
+    max_expansions: int = 1000
+    ) -> Tuple[List[_S], List[_A]]:
+    """A generic heuristic search implementation.
+
+    Depending on get_priority, can implement A*, GBFS, or UCS.
+
+    If no goal is found, returns the node with the best priority.
+    """
+    queue: List[Tuple[Any, int, _HeuristicSearchNode[_S, _A]]] = []
+    frontier = set()
+    explored = set()
+
+    root_node: _HeuristicSearchNode[_S, _A] = _HeuristicSearchNode(
+        initial_state, 0, 0)
+    root_priority = get_priority(root_node)
+    best_node = root_node
+    best_node_priority = root_priority
+    tiebreak = itertools.count()
+    hq.heappush(queue, (root_priority, next(tiebreak), root_node))
+    frontier.add(initial_state)
+    num_expansions = 0
+
+    while len(queue) > 0 and num_expansions < max_expansions:
+        _, _, node = hq.heappop(queue)
+        # If the goal holds, return.
+        if check_goal(node.state):
+            return _finish_plan(node)
+        # Add node state to explored set.
+        explored.add(node.state)
+        num_expansions += 1
+        # Generate successors.
+        for action, child_state, cost in get_successors(node.state):
+            # If the state is already in explored or frontier, don't bother.
+            if child_state in frontier | explored:
```

Comment:
Why do we need both a frontier and an explored? We should be able to just add to explored here when adding a new node, and take out `# Add node state to explored set.` above? Plus the frontier is weird anyway, right, since nothing is being removed from it? (Maybe I'm misunderstanding what the frontier represents.)

Reply:
yeah it's a long and boring story why I had this. I removed and hopefully fixed